File: examples/cpp/stats_example.cpp
#include <cstdint>
#include <iostream>
#include <memory>
#include <string>

#include "client/cpp/payload_manager_client.h"
#include "otel_tracer.hpp"
#include "payload/manager/v1.hpp"
#include "traced_channel.hpp"

int main(int argc, char** argv) {
  // argv[1]: server endpoint  (default localhost:50051)
  // argv[2]: OTLP gRPC endpoint (default localhost:4317, empty to disable)
  const std::string target  = argc > 1 ? argv[1] : "localhost:50051";
  const std::string otlp_ep = argc > 2 ? argv[2] : "localhost:4317";

  OtelInit(otlp_ep, "cpp-examples");
  auto                                    channel = StartSpanAndMakeChannel(target, "stats_example");
  payload::manager::client::PayloadClient client(channel);

  // Stats returns a tier-wise summary of payload counts and byte usage.
  payload::manager::v1::StatsRequest request;
  auto                               result = client.Stats(request);
  if (!result.ok()) {
    std::cerr << "Stats RPC failed: " << result.status().ToString() << '\n';
    return 1;
  }

  const auto& stats = result.ValueOrDie();
  std::cout << "Payload Manager stats for " << target << '\n';
  std::cout << "payload counts: gpu=" << stats.payloads_gpu() << ", ram=" << stats.payloads_ram() << ", disk=" << stats.payloads_disk() << '\n';
  std::cout << "bytes: gpu=" << stats.bytes_gpu() << ", ram=" << stats.bytes_ram() << ", disk=" << stats.bytes_disk() << '\n';

  OtelEndSpan();
  OtelShutdown();
  return 0;
}

File: examples/python/stats_example.py
#!/usr/bin/env python3

import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO_ROOT / "client/python"))

from payload_manager_client import PayloadClient, make_channel
from payload.manager.admin.v1 import stats_pb2


def main() -> int:
    # Optional endpoint argument keeps this example flexible across local and
    # remote deployments.
    target = sys.argv[1] if len(sys.argv) > 1 else "localhost:50051"
    client = PayloadClient(make_channel(target))

    # Stats is a lightweight admin call that summarizes payload counts and
    # allocated bytes per storage tier.
    stats = client.Stats(stats_pb2.StatsRequest())
    print(f"Payload Manager stats for {target}")
    print(
        "payload counts: "
        f"gpu={stats.payloads_gpu}, ram={stats.payloads_ram}, disk_hot={stats.payloads_disk_hot}"
    )
    print(f"bytes: gpu={stats.bytes_gpu}, ram={stats.bytes_ram}, disk_hot={stats.bytes_disk_hot}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
